RadekP2 tasks

radekpetruska opened the pull request:
- Modify List.jsx to render "todo" list.
- Add new components that allow to add/update/delete list items.

radekpetruska commented:
All bullet points should be reflected.

Suzii commented:
Good job with the changes :) 
I don't think it is a good idea to pass only test as a string to ListItemContainer, though. Object with text property was better as it is easy to extend with more properties (as was mentioned in the Task 1 description).
If you decide to change it, have a look at the React.PropTypes.shape({ ... }) -- cool stuff ;)

Another thing considering Draft's coding style - we use static property declared inside of the class for defining propTypes, instead of specifying them outside, e.g:
```
class SomeComponent extends React.Component {
  static propTypes = {
      projectId: PropTypes.string.isRequired,
  };

  constructor() { ... }

  // other methods
}
```
But of course, your solution is equivalent and it is just matter of internal team guidelines which one should be prefered. :)

radekpetruska commented:
Let's do it in Draft way.

Suzii commented:
Great :) (Y) 
Note: I will leave this pull request open so that you can submit other tasks as well

radekpetruska commented:
That should be all for the first attempt to implement task #2.
I've kept states in child components since I didn't see any sense to move them to the root. Just tell me if I overlooked something. Specifically:
- NewItem.jsx - current text in text box
- ListItemEditableText.jsx - current text in text box
- ListItemContainer - currently rendered mode

Suzii commented:
As for the "all state in root component"... Generally it's better to have all the state at one place and propagate down via props (or once you have flux/redix it's even more starightforward). In case of this task, I would surely leave NewItem.jsx as it is right now. The question is whether or not to move the edited value up. I am not saying your solution is bad, you can leave it as is, since it does not really make sense in this small example. But if you had an item with more properties that could be changed or its values could be chenged "from outside", or its current value changes something else in the app -- then you should consider moving it to root (or edux store) :)

radekpetruska commented:
Task \#3:
Application state is now completely stored in Redux store => your previous suggestions were reflected in code:
- _updateState() is gone
- Both edited value and rendered mode is driven by props retrieved from Redux store

Suzii commented on onboarding-task/src/actions/actionTypes.js:
```diff
@@ -0,0 +1,10 @@
+const ITEM_CHECKIN = 'ITEM_CHECKIN';
+const ITEM_DELETE = 'ITEM_DELETE';
+
+export {NEWITEM_UPDATE, ITEM_INSERT, ITEM_CHECKOUT, ITEM_UPDATE, ITEM_REVERT, ITEM_CHECKIN, ITEM_DELETE};
```
You can use export default if you grouped the constants into an object. Alternatively use export const for each constant you want to export. But I would say one default export is preffered :)

Suzii commented on onboarding-task/src/reducers/itemActionsReducer.js:
```diff
@@ -0,0 +1,109 @@
+const insertItem = (state, itemText) => {
+  const newItem = Immutable.Map({
+    [TEXT]: itemText,
+    [ID]: createGuid(),
```
You should never generate stuff in reducers. Reducer MUST be a pure function without any side effects and therefore must never call API or do things such Math.Random() or new Date(). Move either item as a whole to action creator or at least pass newlu generated ID with text to be added as payload of the action.

Suzii commented on onboarding-task/src/reducers/itemActionsReducer.js:
```diff
@@ -0,0 +1,109 @@
+};
+
+const itemActionsReducer = (previousStoreState = Immutable.Map(), action) => {
+  switch (action.type) {
```
I would suggest you try appling reducer composition even on this level. State is a Map and your payload should always contain ID of an item (key in Map). Therefore, you can simplyfy other methods to take only the relevent item istead of whole Map where applicable (-> I think only insert and delete would have to be handled separately).

Suzii commented on onboarding-task/src/reducers/combinedReducer.js:
```diff
@@ -0,0 +1,10 @@
+import {combineReducers} from 'redux';
+import ItemActionsReducer from './itemActionsReducer';
```
Why UpperCamelCase?

Suzii commented on onboarding-task/src/components/NewItem.jsx:
```diff
@@ -1,43 +1,41 @@
 }
 
-export default NewItem;
+const mapStateToProps = (dispatch, ownProps) => {
```
I guess this should be renamed to 'mapDispatchToProps'.

Suzii commented on onboarding-task/src/components/NewItem.jsx:
```diff
@@ -1,43 +1,41 @@
 class NewItem extends React.Component {
   static propTypes = {
-    onInsert: React.PropTypes.func.isRequired
+    text: React.PropTypes.string.isRequired
```
Why not updateText() and insertItem() functions as well?

Suzii commented on onboarding-task/src/components/ListItemEditableText.jsx:
```diff
@@ -1,46 +1,50 @@
+import {TEXT} from '../descriptors/itemProperties';
 
 class ListItemEditableText extends React.Component {
   static propTypes = {
```
Specify callback functions which are passed as props as well.

Suzii commented on onboarding-task/src/components/ListItemContainer.jsx:
```diff
@@ -1,52 +1,27 @@
+import {TEXT, IS_CHECKED_OUT} from '../descriptors/itemProperties';
 
 class ListItemContainer extends React.Component {
   static propTypes = {
```
Add all calback functions as well.

radekpetruska commented:
ad) descriptors: I haven't seen this before but I just wanted to have some constants to avoid typos since item properties are used on a lot of places + it would be easier to rename properties if needed.

Suzii commented on onboarding-task/src/App.jsx:
```diff
@@ -1,35 +1,45 @@
+    const logger = createLogger();
+    const storeEnhancers = applyMiddleware(logger);
+    const store = createStore(reducer, storeEnhancers);
+
```
It is always good to avoid instantiation of objects especially as important as store object in `render()` method. Render is called many times in component lifecycle and therefore new instance will be created every time. You can argue that `App.jsx` should be rendered exactly once but it is a good practice to place this stuff outside anyway. Above class definition after all imports will do :)

Suzii commented on onboarding-task/src/actions/actionTypes.js:
```diff
@@ -0,0 +1,11 @@
+class actionTypes {
```
Why creating a class where simple object literal suffices?

Suzii commented on onboarding-task/src/components/ListItemEditableText.jsx:
```diff
@@ -0,0 +1,54 @@
+}
+
+const mapDispatchToProps = (dispatch, ownProps) => {
+  return {
```
I am not really sure whether this approach is better to sending parameters explicitly from Component. I mean, component should be as much reusable as possible on the other hand, why should redux container be responsible for extracting input value from event object? Personally, I would not use own props at all but that's just my opinion, you don't have to agree and I believe both approaches have their pros and cons.

Suzii commented on onboarding-task/src/components/ListItemEditableText.jsx:
```diff
@@ -0,0 +1,54 @@
+      const action = updateItem(ownProps.item, event.target.value);
+      dispatch(action);
+    },
+    confirmChanges: text => {
```
unused parameter?

Suzii commented on onboarding-task/src/reducers/itemActionsReducer.js:
```diff
@@ -0,0 +1,104 @@
+
+const updateItem = (itemToUpdate, newText) => {
+  if (!itemToUpdate.get(itemProperties.IS_CHECKED_OUT)) {
+    throw new Error('Cannot update checked in item.');
```
Not really sure if throwing in reducer is a good idea. See: https://github.com/reactjs/redux/issues/1960... you can always move this check to actionCreator. I do not have much experience with this, but it feels bad and seems like a side effect from reducer. Let me know your opinion.

radekpetruska commented:
1. I thought so but I was not sure since the render was called only once :smile: 
2. Because C# :smile: 
3. Since I have no previous experience, I changed it so it does not use ownProps. I liked the previous solution just because it was shorter in jsx
4. Resolved in 3)
5. I previously put it here since there is nothing that can stop me from sending actions created by hand (with checked in item) + it helped me to track bugs in unfinished implementation. But since reducers are not expected to throw errors I moved the check to action creator. There is one little change though: according to the article, it's not good to throw an error anyway, so it only log message to console and continue in standard flow.

Suzii commented on onboarding-task/test/reducers/insertFormActionsReducer.test.js:
```diff
@@ -0,0 +1,27 @@
+    const state = insertFormActionsReducer(originalState, action);
+
+    expect(originalState.get(properties.TEXT)).toBe('pasted text');
+    expect(state.get(properties.TEXT)).toBe('updated text');
```
You may also be interested in checking that no other properties of the item were changed.

Suzii commented on onboarding-task/test/testUtils/initialStateGenerator.js:
```diff
@@ -0,0 +1,17 @@
+const defaultItemId = '00-00';
+const defaultItemText = 'default item';
+
+export const createInitialState = () => {
```
Exporting one default object at the end of the file would be preferred.

Suzii commented on onboarding-task/test/testUtils/checkItem.js:
```diff
@@ -0,0 +1,10 @@
+import itemProperties from '../../src/descriptors/itemProperties';
+
+const checkItem = (itemToCheck, text, id, isCheckedOut, originalText) => {
```
This is a bit of magic, too many parameters where the order is super-important. It might be nice to create something more complex and dumb-proof, like creating closure with `item` and returning an object with `ceckXY(xy)` functions wich would return the outer function and might be called in fluent syntax. In case of a failing check, an error could be thrown? I cannot think of anything simpler right now, we can talk in person about other possible options :)
